libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Observers/CycleDetectionObserver.cs

```diff
@@ -0,0 +1,149 @@
+                    result = cache[hashCode] as T;
+                }
+
+                // If we don't have a cached value for this hash code, we send null as the value.
+                // Pass 2 will exist with the purpose of filling in these nulled values with the 
+                // references discovered and cached during pass 1
+                else
+                {
+                    result = null;
+                }
+
+                return true;
+            }
+
+            if (CycleDetectionPass == CycleDetectionPasses.PassTwo) 
+            {
+                // If we already visited this item in pass 2 means we found a loop.
+                // Since in pass 1 we should have filled the cache with the missing objects,
+                // now we just bring the items from pass 1 to stitch the full object together
+                if (visitedPassTwo.Contains(hashCode))
+                {
+                    // We found a loop and we have the final value in the cache. Return that value.
+                    if (cache.ContainsKey(hashCode))
+                    {
+                        result = cache[hashCode] as T;
+                    }
+
+                    // Even if the value was no in the cache, we set as null since we don't want
+                    // to have an infinite loop in pass 2 either.
+                    else
+                    {
+                        result = null;
+                    }
+
+                    return true;
+                }
+
+                visitedPassTwo.Add(hashCode);
+            }
+
+            visitedPassOne.Add(hashCode);
+            result = null;
+            return false;
+        }
+
+        /// <inheritdoc/>
+        public bool OnAfterLoadToken<T>(JToken token, T loaded, out T result)
+            where T : class
+        {
+            // The deep hash code provides a pseudo-unique id for a given token
+            var hashCode = Hash<T>(token);
+
+            // In pass 1, after we build an object, we add it to the cache.
+            if (CycleDetectionPass == CycleDetectionPasses.PassOne) 
+            {
+                cache[hashCode] = loaded;
+                result = null;
+                return false;
+            }
+
+            // In pass 2, we track the objects that we visited to avoid
+            // infinite loops in pass 2. If we have a value in the cache,
+            // we prefer the cache value.
+            else
+            {
+                visitedPassTwo.Add(hashCode);
+                
+                result = null;
+                return false;
+            }
+        }
+
+        private static int Hash<T>(JToken jToken)
+        {
+            // The same json may resolve to two types. The cache key should include
+            // type information.
+            return CombineHashCodes(new[] { jToken.GetHashCode(), typeof(T).GetHashCode() });
+        }
+
+        private static int CombineHashCodes(IEnumerable<int> hashCodes)
```

Comment:
>CombineHashCodes [](start = 27, length = 16)

Does cycledetection always happen in 1 process on 1 machine?

Reply:
Cycle detection is ready to be done as concurrently as json loading - imagine a multi-tenant bot kind of scenario, which is why I did a lot of effort that the orchestration does not depend on global state, or on member variables (i.e. the pass is a method variable, and so is the cycle detector)

I don't want to implement my own hash code, given that these 2 classes already implemented it (JToken and Type). The json objects are in JToken format, and calculating a deep hash is error prone and computationally costly, so I prefer using JToken.GetHashCode. 

The only thing I'm doing here is combining 2 hash codes, which is done with prime numbers and xors. I got the code from the internal string implementation of combine hash codes, since the framework exposes `HashCode.Combine` but not to all the versions of dotnet that we support, which means I cannot use that one. Moving to string and re-hashing will always be less efficient imo. I can move to a separate class but this one is not about strings really. Maybe I can move to a CombinedHash utility?